In [1]:
import warnings
warnings.filterwarnings("ignore")
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

## Exploratory Data Analysis 

In [38]:
#load data
data=pd.read_csv(r"Melbourne_housing_FULL.csv")
data.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/09/2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  4/02/2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt         CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN  Yarra City Council  -37.8014    144.9958  Northern Metropolitan   
1        NaN  Yarra City Council  -37.7996    144.9984  Northern Metropolitan   
2     1900.0  Yarra City Council  -37.8079    144.9934  Northern Metropolitan   
3        NaN  Yarra City Council  -37.8114    145.0116  Northern Metropolitan   
4     1900.0  Yarra City Council  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [39]:
data.nunique()

Suburb             351
Address          34009
Rooms               12
Type                 3
Price             2871
Method               9
SellerG            388
Date                78
Distance           215
Postcode           211
Bedroom2            15
Bathroom            11
Car                 15
Landsize          1684
BuildingArea       740
YearBuilt          160
CouncilArea         33
Lattitude        13402
Longtitude       14524
Regionname           8
Propertycount      342
dtype: int64

In [40]:
# let's use limited columns which makes more sense for serving our purpose
cols_to_use = ['Suburb', 'Rooms', 'Type', 'Method', 'SellerG', 'Regionname', 'Propertycount', 
               'Distance', 'CouncilArea', 'Bedroom2', 'Bathroom', 'Car', 'Landsize', 'BuildingArea', 'Price']
data = data[cols_to_use]

In [41]:
data.head()

Suburb  Rooms Type Method SellerG             Regionname  \
0  Abbotsford      2    h     SS  Jellis  Northern Metropolitan   
1  Abbotsford      2    h      S  Biggin  Northern Metropolitan   
2  Abbotsford      2    h      S  Biggin  Northern Metropolitan   
3  Abbotsford      3    u     VB  Rounds  Northern Metropolitan   
4  Abbotsford      3    h     SP  Biggin  Northern Metropolitan   

   Propertycount  Distance         CouncilArea  Bedroom2  Bathroom  Car  \
0         4019.0       2.5  Yarra City Council       2.0       1.0  1.0   
1         4019.0       2.5  Yarra City Council       2.0       1.0  1.0   
2         4019.0       2.5  Yarra City Council       2.0       1.0  0.0   
3         4019.0       2.5  Yarra City Council       3.0       2.0  1.0   
4         4019.0       2.5  Yarra City Council       3.0       2.0  0.0   

   Landsize  BuildingArea      Price  
0     126.0           NaN        NaN  
1     202.0           NaN  1480000.0  
2     156.0          79.0  1035000.0  
3       0.0           NaN        NaN  
4     134.0         150.0  1465000.0

In [42]:
data.shape

(34857, 15)

In [43]:
# Checking Null Values
data.isna().sum()

Suburb               0
Rooms                0
Type                 0
Method               0
SellerG              0
Regionname           3
Propertycount        3
Distance             1
CouncilArea          3
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
Price             7610
dtype: int64

#### Handling Missing values

In [44]:
# Some feature's missing values can be treated as zero (another class for NA values or absence of that feature)
# like 0 for Propertycount, Bedroom2 will refer to other class of NA values
# like 0 for Car feature will mean that there's no car parking feature with house
cols_to_fill_zero = ['Propertycount', 'Distance', 'Bedroom2', 'Bathroom', 'Car']
data[cols_to_fill_zero] = data[cols_to_fill_zero].fillna(0)

# other continuous features can be imputed with mean for faster results since our focus is on Reducing overfitting
# using Lasso and Ridge Regression
data['Landsize'] = data['Landsize'].fillna(data.Landsize.mean())
data['BuildingArea'] = data['BuildingArea'].fillna(data.BuildingArea.mean())

#### Drop NA values of Price, since it's our predictive variable we won't impute it

In [45]:
data.dropna(inplace=True)

In [46]:
data.shape

(27244, 15)

#### Let's one hot encode the categorical features

In [47]:
data = pd.get_dummies(data, drop_first=True)

In [48]:
data.head()

Rooms  Propertycount  Distance  Bedroom2  Bathroom  Car  Landsize  \
1      2         4019.0       2.5       2.0       1.0  1.0     202.0   
2      2         4019.0       2.5       2.0       1.0  0.0     156.0   
4      3         4019.0       2.5       3.0       2.0  0.0     134.0   
5      3         4019.0       2.5       3.0       2.0  1.0      94.0   
6      4         4019.0       2.5       3.0       1.0  2.0     120.0   

   BuildingArea      Price  Suburb_Aberfeldie  ...  \
1      160.2564  1480000.0                  0  ...   
2       79.0000  1035000.0                  0  ...   
4      150.0000  1465000.0                  0  ...   
5      160.2564   850000.0                  0  ...   
6      142.0000  1600000.0                  0  ...   

   CouncilArea_Moorabool Shire Council  CouncilArea_Moreland City Council  \
1                                    0                                  0   
2                                    0                                  0   
4                                    0                                  0   
5                                    0                                  0   
6                                    0                                  0   

   CouncilArea_Nillumbik Shire Council  CouncilArea_Port Phillip City Council  \
1                                    0                                      0   
2                                    0                                      0   
4                                    0                                      0   
5                                    0                                      0   
6                                    0                                      0   

   CouncilArea_Stonnington City Council  CouncilArea_Whitehorse City Council  \
1                                     0                                    0   
2                                     0                                    0   
4                                     0                                    0   
5                                     0                                    0   
6                                     0                                    0   

   CouncilArea_Whittlesea City Council  CouncilArea_Wyndham City Council  \
1                                    0                                 0   
2                                    0                                 0   
4                                    0                                 0   
5                                    0                                 0   
6                                    0                                 0   

   CouncilArea_Yarra City Council  CouncilArea_Yarra Ranges Shire Council  
1                               1                                       0  
2                               1                                       0  
4                               1                                       0  
5                               1                                       0  
6                               1                                       0  

[5 rows x 745 columns]

In [49]:
X = data.drop('Price', axis=1)
y = data['Price']

In [50]:
X

Rooms  Propertycount  Distance  Bedroom2  Bathroom  Car    Landsize  \
1          2         4019.0       2.5       2.0       1.0  1.0  202.000000   
2          2         4019.0       2.5       2.0       1.0  0.0  156.000000   
4          3         4019.0       2.5       3.0       2.0  0.0  134.000000   
5          3         4019.0       2.5       3.0       2.0  1.0   94.000000   
6          4         4019.0       2.5       3.0       1.0  2.0  120.000000   
...      ...            ...       ...       ...       ...  ...         ...   
34852      4         6543.0       6.3       4.0       1.0  3.0  593.000000   
34853      2         6543.0       6.3       2.0       2.0  1.0   98.000000   
34854      2         6543.0       6.3       2.0       1.0  2.0  220.000000   
34855      3         6543.0       6.3       0.0       0.0  0.0  593.598993   
34856      2         6543.0       6.3       2.0       1.0  0.0  250.000000   

       BuildingArea  Suburb_Aberfeldie  Suburb_Airport West  ...  \
1          160.2564                  0                    0  ...   
2           79.0000                  0                    0  ...   
4          150.0000                  0                    0  ...   
5          160.2564                  0                    0  ...   
6          142.0000                  0                    0  ...   
...             ...                ...                  ...  ...   
34852      160.2564                  0                    0  ...   
34853      104.0000                  0                    0  ...   
34854      120.0000                  0                    0  ...   
34855      160.2564                  0                    0  ...   
34856      103.0000                  0                    0  ...   

       CouncilArea_Moorabool Shire Council  CouncilArea_Moreland City Council  \
1                                        0                                  0   
2                                        0                                  0   
4                                        0                                  0   
5                                        0                                  0   
6                                        0                                  0   
...                                    ...                                ...   
34852                                    0                                  0   
34853                                    0                                  0   
34854                                    0                                  0   
34855                                    0                                  0   
34856                                    0                                  0   

       CouncilArea_Nillumbik Shire Council  \
1                                        0   
2                                        0   
4                                        0   
5                                        0   
6                                        0   
...                                    ...   
34852                                    0   
34853                                    0   
34854                                    0   
34855                                    0   
34856                                    0   

       CouncilArea_Port Phillip City Council  \
1                                          0   
2                                          0   
4                                          0   
5                                          0   
6                                          0   
...                                      ...   
34852                                      0   
34853                                      0   
34854                                      0   
34855                                      0   
34856                                      0   

       CouncilArea_Stonnington City Council  \
1                                         0   
2                                         0   
4        

### Multiple linear regression - least squares fitting

In [55]:
#splitting data for training and testing
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3,random_state=2) #0.67 data will be for training.

In [56]:
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X_train,y_train)
y_predlr=lin_reg.predict(X_test)

In [57]:
from sklearn.metrics import mean_squared_error #calculating MSE
#Training Error
y_pred_train=lin_reg.predict(X_train)
LR_Training_Error=mean_squared_error(y_pred_train,y_train)
print(f"LR_Training_Error for Train:{LR_Training_Error}")

LR_Error=mean_squared_error(y_predlr,y_test)
print(f"LR_Error for test:{LR_Error}")

trainAccuracy= lin_reg.score(X_train,y_train)
testAccuracy= lin_reg.score(X_test,y_test)

print(f"Train Accuarcy : {trainAccuracy}")
print(f"Test Accuarcy : {testAccuracy}")

LR_Training_Error for Train:132454004507.99455
LR_Error for test:342275472104.14484
Train Accuarcy : 0.6827792395792723
Test Accuarcy : 0.1385368316153881


##### Here training score is 68% but test score is 13.85% which is very low

#### Using Ridge (L2 Regularized) Regression Model

In [59]:
from sklearn.linear_model import Ridge
ridge_reg= Ridge(alpha=50, max_iter=100, tol=0.1)
ridge_reg.fit(X_train, y_train)

Ridge(alpha=50, max_iter=100, tol=0.1)

In [60]:
ridge_reg.score(X_test,y_test)

0.6670848945194958

In [62]:
ridge_reg.score(X_train, y_train)

0.6622376739684328

In [67]:
#to get the best parameter for the model
#ridge_reg.coef_

#### Using Lasso (L1 Regularized) Regression Model

In [69]:
from sklearn import linear_model
lasso_reg = linear_model.Lasso(alpha=50, max_iter=100, tol=0.1)
lasso_reg.fit(X_train, y_train)

Lasso(alpha=50, max_iter=100, tol=0.1)

In [70]:
lasso_reg.score(X_train, y_train)

0.6766985624766824

In [71]:
lasso_reg.score(X_test,y_test)

0.6636111369404489

In [74]:
from sklearn.feature_selection import SelectFromModel

In [85]:
model = SelectFromModel(lasso_reg, prefit=True)
X_new = model.transform(X_train)
X_new.shape

(19070, 348)

In [86]:
X_train.shape

(19070, 744)

In [87]:
X_train.columns

Index(['Rooms', 'Propertycount', 'Distance', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'Suburb_Aberfeldie', 'Suburb_Airport West',
       ...
       'CouncilArea_Moorabool Shire Council',
       'CouncilArea_Moreland City Council',
       'CouncilArea_Nillumbik Shire Council',
       'CouncilArea_Port Phillip City Council',
       'CouncilArea_Stonnington City Council',
       'CouncilArea_Whitehorse City Council',
       'CouncilArea_Whittlesea City Council',
       'CouncilArea_Wyndham City Council', 'CouncilArea_Yarra City Council',
       'CouncilArea_Yarra Ranges Shire Council'],
      dtype='object', length=744)

In [83]:
selected_feat = X_train.columns[(model.get_support())]
print('total features: {}'.format((X_train.shape[1])))
print('selected features: {}'.format(len(selected_feat)))
print('features with coefficients shrank to zero: {}'.format(
      np.sum(model.estimator_.coef_ == 0)))

total features: 744
selected features: 348


AttributeError: 'SelectFromModel' object has no attribute 'estimator_'

In [89]:
model.estimator_

AttributeError: 'SelectFromModel' object has no attribute 'estimator_'

In [96]:
from sklearn import linear_model
lasso_reg = linear_model.Lasso(alpha=50, max_iter=100, tol=0.1)

model = SelectFromModel(lasso_reg)
model.fit(X_train, y_train)
selected_feat = X_train.columns[(model.get_support())]
print('total features: {}'.format((X_train.shape[1])))
print('selected features: {}'.format(len(selected_feat)))
print('features with coefficients shrank to zero: {}'.format(
      np.sum(model.estimator_.coef_ == 0)))

total features: 744
selected features: 348
features with coefficients shrank to zero: 396


In [95]:
print('features with coefficients shrank to zero: {}'.format(
      np.sum(sfm.estimator_.coef_ == 0)))

features with coefficients shrank to zero: 396
